flink-mesos/src/main/java/org/apache/flink/mesos/util/MesosArtifactServer.java

```diff
@@ -0,0 +1,304 @@
+import static io.netty.handler.codec.http.HttpResponseStatus.METHOD_NOT_ALLOWED;
+import static io.netty.handler.codec.http.HttpResponseStatus.NOT_FOUND;
+import static io.netty.handler.codec.http.HttpResponseStatus.OK;
+import static io.netty.handler.codec.http.HttpVersion.HTTP_1_1;
+
+
+/**
+ * A generic Mesos artifact server, designed specifically for use by the Mesos Fetcher.
+ *
+ * More information:
+ * http://mesos.apache.org/documentation/latest/fetcher/
+ * http://mesos.apache.org/documentation/latest/fetcher-cache-internals/
+ */
+public class MesosArtifactServer {
+
+	private static final Logger LOG = LoggerFactory.getLogger(MesosArtifactServer.class);
+
+	private final Router router;
+
+	private ServerBootstrap bootstrap;
+
+	private Channel serverChannel;
+
+	private URL baseURL;
+
+	public MesosArtifactServer(String sessionID, String serverHostname, int configuredPort) throws Exception {
+		if (configuredPort < 0 || configuredPort > 0xFFFF) {
+			throw new IllegalArgumentException("File server port is invalid: " + configuredPort);
+		}
+
+		router = new Router();
+
+		ChannelInitializer<SocketChannel> initializer = new ChannelInitializer<SocketChannel>() {
+
+			@Override
+			protected void initChannel(SocketChannel ch) {
+				Handler handler = new Handler(router);
+
+				ch.pipeline()
+					.addLast(new HttpServerCodec())
+					.addLast(handler.name(), handler)
+					.addLast(new UnknownFileHandler());
+			}
+		};
+
+		NioEventLoopGroup bossGroup   = new NioEventLoopGroup(1);
+		NioEventLoopGroup workerGroup = new NioEventLoopGroup();
+
+		this.bootstrap = new ServerBootstrap();
+		this.bootstrap
+			.group(bossGroup, workerGroup)
+			.channel(NioServerSocketChannel.class)
+			.childHandler(initializer);
+
+		Channel ch = this.bootstrap.bind(serverHostname, configuredPort).sync().channel();
+		this.serverChannel = ch;
+
+		InetSocketAddress bindAddress = (InetSocketAddress) ch.localAddress();
+		String address = bindAddress.getAddress().getHostAddress();
+		int port = bindAddress.getPort();
+
+		baseURL = new URL("http", serverHostname, port, "/" + sessionID + "/");
+
+		LOG.info("Mesos artifact server listening at " + address + ':' + port);
+	}
+
+	/**
+	 * Get the server port on which the artifact server is listening.
+     */
+	public synchronized int getServerPort() {
+		Channel server = this.serverChannel;
+		if (server != null) {
+			try {
+				return ((InetSocketAddress) server.localAddress()).getPort();
+			} catch (Exception e) {
+				LOG.error("Cannot access local server port", e);
+			}
+		}
+		return -1;
+	}
+
+	/**
+	 * Adds a file to the artifact server.
+	 * @param localFile the local file to serve.
+	 * @param remoteFile the remote path with which to locate the file.
+	 * @return the fully-qualified remote path to the file.
+	 * @throws MalformedURLException if the remote path is invalid.
+     */
+	public synchronized URL addFile(File localFile, String remoteFile) throws MalformedURLException {
```

Comment:
`synchronized` really necessary? Should only be called by a single process, i.e. the mesos master runner